src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1821,171 +1821,378 @@ PhaseStatus Compiler::fgLateCastExpansion()
+        {
+            if (isCastClass)
+            {
+                // Fallback call is only needed for castclass and only to throw InvalidCastException
+                *typeCheckFailed = F_CallHelper_AlwaysThrows;
+
+                // Assume that exceptions are rare
+                *likelihood = 100;
+            }
+            else
+            {
+                // Fallback for isinst simply returns null here
+                *typeCheckFailed = F_ReturnNull;
+            }
+        }
+        else
+        {
+            // 3) Check PGO data
+            LikelyClassMethodRecord likelyClasses[MAX_GDV_TYPE_CHECKS];
+            unsigned                likelyClassCount =
+                getLikelyClasses(likelyClasses, MAX_GDV_TYPE_CHECKS, comp->fgPgoSchema, comp->fgPgoSchemaCount,
+                                 comp->fgPgoData, (int)castHelper->gtCastHelperILOffset);
+
+            if (likelyClassCount != 0)
+            {
+#ifdef DEBUG
+                // Print all the candidates and their likelihoods to the log
+                for (UINT32 i = 0; i < likelyClassCount; i++)
+                {
+                    const char* className = comp->eeGetClassName((CORINFO_CLASS_HANDLE)likelyClasses[i].handle);
+                    JITDUMP("  %u) %p (%s) [likelihood:%u%%]\n", i + 1, likelyClasses[i].handle, className,
+                            likelyClasses[i].likelihood);
+                }
+
+                // Optional stress mode to pick a random known class, rather than
+                // the most likely known class.
+                if (JitConfig.JitRandomGuardedDevirtualization() != 0)
+                {
+                    // Reuse the random inliner's random state.
+                    CLRRandom* const random = comp->impInlineRoot()->m_inlineStrategy->GetRandom(
+                        JitConfig.JitRandomGuardedDevirtualization());
+                    unsigned index = static_cast<unsigned>(random->Next(static_cast<int>(likelyClassCount)));
+
+                    likelyClasses[0].likelihood = 100;
+                    likelyClasses[0].handle     = likelyClasses[index].handle;
+                }
+#endif
+
+                // if there is a dominating candidate with >= 50% likelihood, use it
+                const unsigned likelihoodMinThreshold = 50;
+                if (likelyClasses[0].likelihood < likelihoodMinThreshold)
+                {
+                    JITDUMP("Likely class likelihood is below %u%% - bail out.\n", likelihoodMinThreshold);
+                    return NO_CLASS_HANDLE;
+                }
+
+                *likelihood = likelyClasses[0].likelihood;
+                result      = (CORINFO_CLASS_HANDLE)likelyClasses[0].handle;
+
+                // Validate static profile data
+                if ((comp->info.compCompHnd->getClassAttribs(result) &
+                     (CORINFO_FLG_INTERFACE | CORINFO_FLG_ABSTRACT | CORINFO_FLG_STATIC)) != 0)
```

Comment:
Is `CORINFO_FLG_STATIC` possible on a class? I think it only applies to methods. Static classes in C# are abstract and sealed in IL, I think.

Reply:
Absolutely not needed, no idea why I added it here. this check was initially added to handle stale static pgo